File: app/src/main/java/com/dumuzeyn/mp3player/ExcludedTrack.java
package com.dumuzeyn.mp3player;

/** Persistent user decision to keep one physical audio file out of Voltune. */
final class ExcludedTrack {
    final String identityKey;
    final String sourceId;
    final String documentId;
    final String uri;
    final long fileSize;
    final long lastModified;
    final String fingerprint;

    ExcludedTrack(String identityKey, String sourceId, String documentId, String uri,
            long fileSize, long lastModified, String fingerprint) {
        this.identityKey = identityKey;
        this.sourceId = sourceId == null ? "" : sourceId;
        this.documentId = documentId == null ? "" : documentId;
        this.uri = uri == null ? "" : uri;
        this.fileSize = fileSize;
        this.lastModified = lastModified;
        this.fingerprint = fingerprint == null ? "" : fingerprint;
    }

    static ExcludedTrack from(Track track, TrackOrigin origin) {
        String key = origin == null ? TrackOrigin.uriIdentity(track.uri) : origin.identityKey;
        return new ExcludedTrack(key, origin == null ? "" : origin.sourceId,
                origin == null ? "" : origin.documentId, track.uri, track.fileSize,
                track.lastModified, track.fingerprint);
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/ExcludedTrack.kt
package com.dumuzeyn.mp3player

/** Persistent user decision to keep one physical audio file out of Voltune. */
class ExcludedTrack(
    @JvmField val identityKey: String,
    sourceId: String?,
    documentId: String?,
    uri: String?,
    @JvmField val fileSize: Long,
    @JvmField val lastModified: Long,
    fingerprint: String?,
) {
    @JvmField val sourceId: String = sourceId.orEmpty()
    @JvmField val documentId: String = documentId.orEmpty()
    @JvmField val uri: String = uri.orEmpty()
    @JvmField val fingerprint: String = fingerprint.orEmpty()

    companion object {
        @JvmStatic
        fun from(track: Track, origin: TrackOrigin?): ExcludedTrack = ExcludedTrack(
            origin?.identityKey ?: TrackOrigin.uriIdentity(track.uri),
            origin?.sourceId,
            origin?.documentId,
            track.uri,
            track.fileSize,
            track.lastModified,
            track.fingerprint,
        )
    }
}
